torch/csrc/jit/passes/onnx/peephole.cpp

```diff
@@ -830,6 +830,153 @@ void removeMaxPoolUnusedOutput(Block* b) {
+  }
+  node->removeAllInputs();
+
+  if (node->kind() != onnx::NegativeLogLikelihoodLoss) {
+    bool del = true;
+    for (size_t i = 0; i < node->outputs().size(); ++i) {
+      if (!node->outputs()[i]->uses().empty()) {
+        del = false;
+        break;
+      }
+    }
+    if (del) {
+      node->destroy();
+    }
+  }
+
+  for (Node* n: nodeList) {
+    recursive_del(n);
+  }
+}
+
+Node* createSoftmaxCrossEntropyNode(Block* b, graph_node_list_iterator it, std::vector<Value*> values) {
+  Node* softmaxCrossEntropyNode = b->owningGraph()->create(onnx::SoftmaxCrossEntropyLoss, it->outputs().size());
+  for (size_t i = 0; i < softmaxCrossEntropyNode->outputs().size(); ++i) {
+    softmaxCrossEntropyNode->outputs()[i]->copyMetadata(it->outputs()[i]);
+  }
+  for (Value* value : values) {
+    softmaxCrossEntropyNode->addInput(value);
+  }
+  return softmaxCrossEntropyNode;
+}
+
+// This optimization fuses LogSoftmax and NegativeLogLikelihoodLoss operators into
+// one operator: SoftmaxCrossEntropyLoss.
+static void fuseLogSoftmaxNllLoss(Block* b) {
+  for (auto it = b->nodes().begin(), end = b->nodes().end(); it != end; ++it) {
+    for (auto* child_block : it->blocks()) {
+      fuseLogSoftmaxNllLoss(child_block);
+    }
+    if (it->kind() == onnx::NegativeLogLikelihoodLoss) {
+      auto prev = it->input(0)->node();
+      Node* origLogSoftmaxNode;
+      Node* origNllLossNode;
+      std::vector<Node*> deleteNodes;
+      Node* softmaxCrossEntropyNode = nullptr;
+      // if the input is 2D
+      if (prev->kind() == onnx::LogSoftmax) {
+        origLogSoftmaxNode= it->input(0)->node();
+        origNllLossNode = *it;
+        deleteNodes.push_back(origLogSoftmaxNode);
+      // if the input is 4D
+      } else if (prev->kind() == onnx::Transpose &&
+                 prev->input(0)->node()->kind() == onnx::LogSoftmax) {
+        auto logSoftmaxNode = prev->input(0)->node();
+        origLogSoftmaxNode = logSoftmaxNode->input(0)->node();
+        origNllLossNode = *it;
+        prev->removeAllInputs();
+        logSoftmaxNode->removeAllInputs();
+        deleteNodes.push_back(origLogSoftmaxNode);
+        deleteNodes.push_back(logSoftmaxNode);
+        deleteNodes.push_back(prev);
+      // if the input is 3D or > 4D
+      } else if (prev->kind() == onnx::Reshape &&
+                 prev->input(0)->node()->kind() == onnx::Transpose &&
+                 prev->input(0)->node()->input(0)->node()->kind() == onnx::LogSoftmax) {
+        origLogSoftmaxNode = prev->input(0)->node()->input(0)->node();
+        auto transpose = origLogSoftmaxNode->input(0)->node();
+        origNllLossNode = *it;
+        auto reshape = origNllLossNode->input(1)->node();
+        std::vector<Value*> values{transpose->inputs().at(0), reshape->inputs().at(0)};
+        if (origNllLossNode->inputs().size() == 3) {
+          values.push_back(origNllLossNode->inputs().at(2));
+        }
+        softmaxCrossEntropyNode = createSoftmaxCrossEntropyNode(b, it, values);
```

Comment:
And after removing code that are deleting stuff, this part that creating softmax crossentropy node separately can be refactored.